tests/components/media_player/test_cast.py

```diff
@@ -18,83 +21,221 @@ def cast_mock():
+        signal = MagicMock()
+
+        async_dispatcher_connect(hass, 'cast_discovered', signal)
+        discover_cast('the-service', chromecast)
+        yield from hass.async_block_till_done()
+
+        assert signal.call_count == 0
+
+
+def test_create_cast_device_without_uuid(hass):
+    """Test create a cast device without a UUID."""
+    chromecast = get_fake_chromecast(uuid=None)
+    cast_device = cast._async_create_cast_device(hass, chromecast)
+    assert cast_device is not None
+
+
+def test_create_cast_device_with_uuid(hass):
+    """Test create cast devices with UUID."""
+    added_casts = hass.data[cast.ADDED_CAST_DEVICES_KEY] = {}
+    chromecast = get_fake_chromecast()
+    cast_device = cast._async_create_cast_device(hass, chromecast)
+    assert cast_device is not None
+    assert chromecast.uuid in added_casts
+
+    with patch.object(cast_device, 'async_set_chromecast') as mock_set:
+        assert cast._async_create_cast_device(hass, chromecast) is None
+        assert mock_set.call_count == 0
+
+        chromecast = get_fake_chromecast(host='192.168.178.1')
+        assert cast._async_create_cast_device(hass, chromecast) is None
+        assert mock_set.call_count == 1
+        mock_set.assert_called_once_with(chromecast)
+
+
+@asyncio.coroutine
+def test_normal_chromecast_not_starting_discovery(hass):
+    """Test cast platform not starting discovery when not required."""
+    chromecast = get_fake_chromecast()
+
+    with patch('pychromecast.Chromecast', return_value=chromecast):
+        add_devices = yield from async_setup_cast(hass, {'host': 'host1'})
+        assert add_devices.call_count == 1
+
+        # Same entity twice
+        add_devices = yield from async_setup_cast(hass, {'host': 'host1'})
+        assert add_devices.call_count == 0
+
+        hass.data[cast.ADDED_CAST_DEVICES_KEY] = {}
+        add_devices = yield from async_setup_cast(
+            hass, discovery_info={'host': 'host1', 'port': 8009})
+        assert add_devices.call_count == 1
+
+        hass.data[cast.ADDED_CAST_DEVICES_KEY] = {}
+        add_devices = yield from async_setup_cast(
+            hass, discovery_info={'host': 'host1', 'port': 42})
+        assert add_devices.call_count == 0
+
+
+@asyncio.coroutine
+def test_replay_past_chromecasts(hass):
+    """Test cast platform re-playing past chromecasts when adding new one."""
+    cast_group1 = get_fake_chromecast(host='host1', port=42)
+    cast_group2 = get_fake_chromecast(host='host2', port=42, uuid=UUID(
+        '9462202c-e747-4af5-a66b-7dce0e1ebc09'))
+
+    discover_cast, add_dev1 = yield from async_setup_cast_internal_discovery(
+        hass, discovery_info={'host': 'host1', 'port': 42})
+    discover_cast('service2', cast_group2)
+    yield from hass.async_block_till_done()
+    assert add_dev1.call_count == 0
+
+    discover_cast('service1', cast_group1)
+    yield from hass.async_block_till_done()
+    yield from hass.async_block_till_done()  # having jobs that add jobs
```

Comment:
I'm not sure why I have to do this, but it works 😅 (I suspect it's due to "jobs that add jobs", but really I don't know)